### PR TITLE
DataJoint core module for Python

### PR DESCRIPTION
**Pull request by JHocevar**
The main goal of this PR is to move the python wrapper of the Rust CFFI to a `datajoint_core` submodule inside the `datajoint` package. All references to the FFI should now take place inside this sub module. Additionally, this allows the main `datajoint` package to function closer to how the main [datajoint](https://github.com/datajoint/datajoint-python) package functions, with a single connection and configuration objects. This functionality could be expanded in the future, but this allows us to more easily prepare to merge into the main repo first. 

## Issues Addressed
*  #73 - Main issue addressed by PR, explained in introduction.
* #78 - Added `free_and_decode_string` method to `connection_config` to free strings received from the FFI
* #90 - Refactored how config settings are set using dictionaries keyed by the setting type for generalized encoding and decoding from the FFI

## Future work to be addressed in later PR's
* Fix build script and setup.py to work with multi module setup
* Look into porting over the rest of the settings from python into rust library.
  * Also possibly remove the `database_name` setting as in [datajoint-python](https://github.com/datajoint/datajoint-python) this is used in the [query](https://github.com/datajoint/datajoint-python/blob/72c8fea6aa5b111a1007ada1b534fe1a510979da/datajoint/table.py#L34) rather than the connection uri
* Continue integration with [datajoint-python](https://github.com/datajoint/datajoint-python) to prepare for PR into the main repo. 
  * Replace uses of pymysql with our new sub module and Connection class

### CODE REVIEW

File: datajoint-python/datajoint/__init__.py [hunk 1]
```diff
@@ -1,11 +1,49 @@
...
 
-from .connection import conn, Connection
-from .cursor import Cursor
-from .datajoint_types import DataJointType
-from .settings import Config
-from .table_column_ref import TableColumnRef
-from .table_row import TableRow
-from .table_row_vector import TableRowVector
+__author__ = "DataJoint Contributors"
+__date__ = "November 7, 2020"
```
Comment (Rhijke) [inline datajoint-python/datajoint/__init__.py:18]: 2020 🤔

File: datajoint-python/test.py [hunk 2]
```diff
@@ -2,8 +2,8 @@
 import datajoint as dj
 
 
-connection = dj.conn(host="tutorial-db.datajoint.io", user="<username>",
-                     password="<password>", database_name="jonathan_tutorial", reset=False, use_tls=True)
+connection = dj.conn(host="tutorial-db.datajoint.io", user="jonathan",
+                     password="testpassword123", database_name="jonathan_tutorial", reset=False, use_tls=True)
 
 cursor = connection.fetch_query(
     "select * from mouse where sex = ?", 'M')
```
Comment (Rhijke) [inline datajoint-python/test.py:9]: uh do we want this file merged into the repo or should we just put the sample code in the README

File: datajoint-python/datajoint/connection.py [hunk 3]
```diff
@@ -1,83 +1,44 @@
...
-    if password is not None:
-        connection.config["password"] = password
-    if database_name is not None:
-        connection.config["database_name"] = database_name
-    if database_type is not None:
-        connection.config["database_type"] = database_type
-    connection.connect()
-    return connection
+"""
+This module contains the Connection class that manages the connection to the database,
```
Comment (Rhijke) [inline datajoint-python/datajoint/connection.py:2]: Is the Connection class getting moved here later? I'm not seeing any class here currently...

File: datajoint-python/datajoint/datajoint_core/connection_config.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            self.native[0] = native
+            self.owning = owning
+        else:
+            raise ValueError("invalid type for native pointer")
+
+    def __del__(self):
+        if self.owning:
+            dj_core.connection_settings_free(self.native[0])
+
+    # TODO(jackson-nestelroad): Type checking here for inputs and outputs.
```
Comment (jackson-nestelroad) [inline datajoint-python/datajoint/datajoint_core/connection_config.py:110]: Can remove this comment now, right?

**Comment by jackson-nestelroad** [comment]
I think I like how this is set up. I especially like how the connection settings class is working now.

File: datajoint-python/datajoint/__init__.py (hunk 1 above)
Comment (JHocevar) [reply datajoint-python/datajoint/__init__.py:18]: This file was copied from the `datajoint-python` repo as that is what our top level `datajoint` module is emulating before we merge it in.

File: datajoint-python/test.py (hunk 2 above)
Comment (JHocevar) [reply datajoint-python/test.py:9]: Probably a good idea. I'll remove this last test file for now

File: datajoint-python/datajoint/connection.py (hunk 3 above)
Comment (JHocevar) [reply datajoint-python/datajoint/connection.py:2]: Not at the moment. We decided to use the Connection class inside the `datajoint-core` module that wraps the core library and FFI and not wrap it again. Idea here was to limit the number of times we wrap the same things.

File: datajoint-python/datajoint/datajoint_core/connection_config.py (hunk 4 above)
Comment (JHocevar) [reply datajoint-python/datajoint/datajoint_core/connection_config.py:110]: I thought so, I left it to make sure your original intentions were addressed. I'll go ahead and remove it

File: README.md [hunk 5]
```diff
@@ -12,25 +12,22 @@ Build the rust library
...
+## Using the libarary
+You can run the python from the interactive shell
 
 ```python
 import datajoint as dj
 
-connection = dj.conn("example@email.com", "Username123", "secretPassword", reset=False, use_tls=True)
-connection.raw_query("SELECT STUFF FROM TABLE")
+connection = dj.conn(host="tutorial-db.datajoint.io", user=<username>,
+                     password=<password>, database_name="jonathan_tutorial", reset=False, use_tls=True)
```
Comment (jonathanschroeter) [inline README.md:22]: Do we want to keep the database_name as "jonathan_tutorial"? It's generic for username and password, so just wanted to see if there is a reason for this.

Comment (JHocevar) [reply README.md:22]: Actually, that setting needs to be removed entirely as it is no longer a parameter of `dj.conn`

Comment (JHocevar) [reply README.md]: fixed

File: README.md [hunk 6]
```diff
@@ -12,25 +12,22 @@ Build the rust library
...
-output
-
-```
-Connected to database: Host: example@email.com User: Username123 Password: secretPassword Reset: false use_TLS: true
-Making query from rust library: SELECT STUFF FROM TABLE
-result from query is: 0
-```
-
-You can also test from the python interactive shell
+## Using the libarary
```
Comment (jackson-nestelroad) [inline README.md:15]: Typo.

File: README.md [hunk 7]
```diff
@@ -12,25 +12,22 @@ Build the rust library
...
+connection = dj.conn(host="tutorial-db.datajoint.io", user=<username>,
+                     password=<password>, reset=False, use_tls=True)
+
+cursor = connection.fetch_query("select * from mouse where sex = ?", 'M')
+try:
+    l = list(cursor)
+    for row in l:
+        print(row.to_dict())
+except AssertionError:
+    print('failed :(')
```
Comment (jackson-nestelroad) [inline README.md:30]: Can we change this to print the actual error?

File: datajoint-python/datajoint/datajoint_core/connection_config.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+
+class ConnectionSetting:
+    def __init__(self, getter, setter, ffi_type):
+        self.getter = getter
+        self.setter = setter
+        self.ffi_type = ffi_type
+
+    def set_value(self, native, value):
+        encode_methods = {
```
Comment (jackson-nestelroad) [inline datajoint-python/datajoint/datajoint_core/connection_config.py:53]: Is there a way to make this (and `decode_methods`) static? They clearly do not need to be local for each instance/call to these setters and getters.

File: README.md (hunk 6 above)
Comment (jackson-nestelroad) [reply README.md:15]: Also, this has been moved to /doc/SETUP.md

Comment (JHocevar-UTD) [reply README.md:15]: fixed

Comment (JHocevar) [reply README.md:15]: fixed

File: datajoint-python/datajoint/datajoint_core/connection_config.py (hunk 8 above)
Comment (JHocevar) [reply datajoint-python/datajoint/datajoint_core/connection_config.py:53]: Moved to top level of file to be static with other encoding and decoding dicts.

File: README.md (hunk 7 above)
Comment (JHocevar) [reply README.md:30]: fixed

**Review by jackson-nestelroad — APPROVED** [review]
Thanks!

**Pull request merged**